astro/snippet.vim

```diff
@@ -0,0 +1,7 @@
+let g:ycm_language_server += [
+  \   {
+  \     'name': 'astro',
+  \     'filetypes': [ 'astro' ],
+  \     'cmdline': [  expand( g:ycm_lsp_dir . '/astro/node_modules/@astrojs/language-server/bin/nodeServer.js' ), '--stdio' ],
```

Comment:
This looks fine, but is there anything we could put in `project_root_files`?
In other words, does Astro have some project descriptor file that usually sits at the root of the project?

If so, we should add that, so that YCM can figure out the root of the project better.

Reply:
**Short answer**; I'm not sure, here's the output of `ls -1` after a _mostly_ fresh Astro project initialization;

```
astro.config.mjs
node_modules
package.json
package-lock.json
public
README.md
src
tsconfig.json
```

... But am thinking the `astro.conf.mjs` and `tsconfig.json` files may be important to the Astro Language Server.

**_Slightly_ longer** answer;

- Official [documentation](https://github.com/withastro/language-tools/tree/main/packages/language-server) is a _bit_ sparse as to what else it may want
- :warning: I did spot the [`@astro/ts-plugin`](https://github.com/withastro/language-tools/tree/main/packages/ts-plugin) may be an additional recommended ritual to add to `npm init` tasks
- Configurations for NeoVim's [Mason LSP](https://github.com/williamboman/mason-lspconfig.nvim/blob/f0ce33f4794a2364eb08d09d09380e8b04ec5e6a/lua/mason-lspconfig/server_configurations/astro/init.lua#L6) plugin seem to match requiring only the `typescript.tsdk` directory
- Searching for [`Vim` Issues](https://github.com/withastro/language-tools/issues?q=is%3Aissue+vim) and skimming results didn't expose any additional requirements or suggestions beyond requiring the `typescript.tsdk` directory
- With the advertised configurations `YcmToggleLogs` logs don't show any complaints, warnings, or notifications related to searching-out additional files;
   ```
   2024-09-02 09:45:06,153 - INFO - Completion config: 50, detailing 0 candiates
   2024-09-02 09:45:06,153 - INFO - Completion config: 50, detailing 0 candiates
   2024-09-02 09:45:06,153 - INFO - Completion config: 50, detailing 0 candiates
   2024-09-02 09:45:06,153 - INFO - Completion config: 50, detailing 0 candiates
   127.0.0.1 - - [02/Sep/2024 09:45:06] "GET /ready HTTP/1.1" 200 4
   2024-09-02 09:45:06,164 - INFO - Initializing generic LSP completer with: {'name': 'astro', 'cmdline': ['/home/USER/git/hub/ycm-core/lsp-examples/astro/node_modules/@astrojs/language-server/bin/nodeServer.js', '--stdio'], 'filetypes': ['astro']}
   2024-09-02 09:45:06,164 - INFO - Completion config: 50, detailing 0 candiates
   127.0.0.1 - - [02/Sep/2024 09:45:06] "GET /signature_help_available?subserver=astro HTTP/1.1" 200 23
   127.0.0.1 - - [02/Sep/2024 09:45:06] "POST /event_notification HTTP/1.1" 200 2
   2024-09-02 09:45:06,383 - INFO - Adding buffer identifiers for file: /home/USER/git/hub/ORG/REPO/src/components/example.astro
   127.0.0.1 - - [02/Sep/2024 09:45:06] "POST /receive_messages HTTP/1.1" 200 5
   2024-09-02 09:45:06,389 - INFO - Starting astroCompleter: ['/home/USER/git/hub/ycm-core/lsp-examples/astro/node_modules/@astrojs/language-server/bin/nodeServer.js', '--stdio']
   2024-09-02 09:45:06,393 - INFO - astroCompleter started with PID 10008
   127.0.0.1 - - [02/Sep/2024 09:45:06] "POST /event_notification HTTP/1.1" 200 2
   127.0.0.1 - - [02/Sep/2024 09:45:06] "POST /semantic_completion_available HTTP/1.1" 200 4
   2024-09-02 09:45:10,393 - INFO - astro: Language server requires resolve request
   2024-09-02 09:45:10,393 - INFO - astro: Language server requires sync type of Incremental
   2024-09-02 09:45:10,393 - INFO - astro: Using trigger characters for semantic triggers: .,:,<,",=,/,-,>,+,^,*,(,),#,[,],$,@,{,},',`, 
   2024-09-02 09:45:10,394 - INFO - astro: Using characters for signature triggers: (,,,<
   127.0.0.1 - - [02/Sep/2024 09:45:10] "POST /receive_messages HTTP/1.1" 200 4
   2024-09-02 09:45:10,403 - WARNING - Server reported: Inlay hint refresh is not supported by the client.
   2024-09-02 09:45:10,403 - WARNING - Server reported: Semantic tokens refresh is not supported by the client.
   127.0.0.1 - - [02/Sep/2024 09:45:13] "POST /receive_messages HTTP/1.1" 200 117
   127.0.0.1 - - [02/Sep/2024 09:45:13] "POST /receive_messages HTTP/1.1" 200 117
   127.0.0.1 - - [02/Sep/2024 09:45:23] "POST /receive_messages HTTP/1.1" 200 4
   127.0.0.1 - - [02/Sep/2024 09:45:33] "POST /receive_messages HTTP/1.1" 200 4
   127.0.0.1 - - [02/Sep/2024 09:45:43] "POST /receive_messages HTTP/1.1" 200 4
   127.0.0.1 - - [02/Sep/2024 09:45:44] "POST /debug_info HTTP/1.1" 200 1121
   127.0.0.1 - - [02/Sep/2024 09:45:46] "POST /receive_messages HTTP/1.1" 200 5
   127.0.0.1 - - [02/Sep/2024 09:45:47] "POST /debug_info HTTP/1.1" 200 1121
   127.0.0.1 - - [02/Sep/2024 09:45:49] "POST /debug_info HTTP/1.1" 200 1121
   127.0.0.1 - - [02/Sep/2024 09:45:53] "POST /receive_messages HTTP/1.1" 200 4
   127.0.0.1 - - [02/Sep/2024 09:45:54] "POST /debug_info HTTP/1.1" 200 1121
   127.0.0.1 - - [02/Sep/2024 09:46:03] "POST /receive_messages HTTP/1.1" 200 4
   2024-09-02 09:47:19,276 - INFO - No support for ExecuteCommand command in server for astro
   2024-09-02 09:47:19,277 - INFO - Found codeActionProvider support for command FixIt in astro
   2024-09-02 09:47:19,277 - INFO - Found definitionProvider support for command GoToDefinition in astro
   2024-09-02 09:47:19,277 - INFO - Found definitionProvider support for command GoToDeclaration in astro
   2024-09-02 09:47:19,277 - INFO - Found definitionProvider support for command GoTo in astro
   2024-09-02 09:47:19,277 - INFO - Found typeDefinitionProvider support for command GoToType in astro
   2024-09-02 09:47:19,277 - INFO - Found implementationProvider support for command GoToImplementation in astro
   2024-09-02 09:47:19,277 - INFO - Found referencesProvider support for command GoToReferences in astro
   2024-09-02 09:47:19,277 - INFO - Found renameProvider support for command RefactorRename in astro
   2024-09-02 09:47:19,278 - INFO - Found documentFormattingProvider support for command Format in astro
   2024-09-02 09:47:19,278 - INFO - Found workspaceSymbolProvider support for command GoToSymbol in astro
   2024-09-02 09:47:19,278 - INFO - Found documentSymbolProvider support for command GoToDocumentOutline in astro
   2024-09-02 09:47:19,278 - INFO - Always supporting StopServer for astro
   2024-09-02 09:47:19,278 - INFO - Always supporting RestartServer for astro
   2024-09-02 09:47:19,278 - INFO - Always supporting GoToCallees for astro
   2024-09-02 09:47:19,278 - INFO - Always supporting GoToCallers for astro
   2024-09-02 09:47:19,278 - INFO - Always supporting CallHierarchy for astro
   2024-09-02 09:47:19,278 - INFO - Always supporting ResolveCallHierarchyItem for astro
   2024-09-02 09:47:19,278 - INFO - Always supporting GetHover for astro
   127.0.0.1 - - [02/Sep/2024 09:47:19] "POST /run_completer_command HTTP/1.1" 200 418
   127.0.0.1 - - [02/Sep/2024 09:47:19] "POST /receive_messages HTTP/1.1" 200 4
   ```